odl/trafos/wavelet.py

```diff
@@ -21,9 +21,836 @@
+        flat_coeff[start:stop] = aad.ravel()
+        start = stop
+        stop = start + np.prod(size_list[kk])
+        flat_coeff[start:stop] = ada.ravel()
+        start = stop
+        stop = start + np.prod(size_list[kk])
+        flat_coeff[start:stop] = add.ravel()
+        start = stop
+        stop = start + np.prod(size_list[kk])
+        flat_coeff[start:stop] = daa.ravel()
+        start = stop
+        stop = start + np.prod(size_list[kk])
+        flat_coeff[start:stop] = dad.ravel()
+        start = stop
+        stop = start + np.prod(size_list[kk])
+        flat_coeff[start:stop] = dda.ravel()
+        start = stop
+        stop = start + np.prod(size_list[kk])
+        flat_coeff[start:stop] = ddd.ravel()
+        start = stop
+
+    return flat_coeff
+
+
+def array_to_pywt_coeff3d(coeff, size_list, nscales):
+    """Convert a flat array into a pywt coefficient list.
+
+    For multilevel 3D discrete wavelet transform
+
+    Parameters
+    ----------
+    coeff :  :class:`DiscreteLp.Vector`
+        A flat coefficient vector containing the approximation,
+        and detail coefficients in the following order
+        [aaaN, aadN, adaN, addN, daaN, dadN, ddaN, dddN, ...
+        aad1, ada1, add1, daa1, dad1, dda1, ddd1]
+
+    size_list : list
+       A list of coefficient sizes such that,
+
+       size_list[0] = size of approximation coefficients at the coarsest level,
+
+       size_list[1] = size of the detailed coefficients at the coarsest level,
+
+       size_list[N] = size of the detailed coefficients at the finest level,
+
+       size_list[N+1] = size of original image,
+
+       N = nscales
+
+    nscales : int
+        Number of scales in the coefficient list
+
+    Returns
+    -------
+    :attr:`coeff_list` : `list`
+        A list of coefficient organized in the following way
+        [aaaN, (aadN, adaN, addN, daaN, dadN, ddaN, dddN), ...
+        (aad1, ada1, add1, daa1, dad1, dda1, ddd1)].
+        The appreviations refer to,
+
+        aaa = approx. on 1st dim, approx. on 2nd dim, approx. on 3rd dim,
+
+        aad = approx. on 1st dim, approx. on 2nd dim, detail on 3rd dim,
+
+        ada = approx. on 1st dim, detail on 3nd dim, approx. on 3rd dim,
+
+        add = approx. on 1st dim, detail on 3nd dim, detail on 3rd dim,
+
+        daa = detail on 1st dim, approx. on 2nd dim, approx. on 3rd dim,
+
+        dad = detail on 1st dim, approx. on 2nd dim, detail on 3rd dim,
+
+        dda = detail on 1st dim, detail on 2nd dim, approx. on 3rd dim,
+
+        ddd = detail on 1st dim, detail on 2nd dim, detail on 3rd dim,
+
+        N = the number of scaling levels
+
+    See also
+    --------
+    pywt : http://www.pybytes.com/pywavelets/
+    """
+
+    size1 = np.prod(size_list[0])
+    approx_flat = coeff[0:size1]
+    approx = np.asarray(approx_flat).reshape(size_list[0])
+    kk = 1
+    coeff_list = []
+    coeff_list.append(approx)
+
+    while kk <= nscales:
```

Comment:
Perhaps try to violate [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) less:

```python
while kk <= nscales:
    detail_shape = size_list[kk]
    size2 = np.prod(size_list[kk])
    for i in range(8):
        flat = coeff[size1+(i-1)*size2:size1 + i*size1]
        details += [np.asarray(dda_flat).reshape(detail_shape)]
        
    coeff_list.append(details)
    kk = kk + 1
    size1 = size1 + 7*size2
```